File: python/order_workflow.py
import logging
from datetime import timedelta
import asyncio

from temporalio import workflow

from activities import OrderActivities
from shared_objects import OrderInput, OrderOutput

logging.basicConfig(level=logging.INFO)


@workflow.defn(name="OrderWorkflowHappyPath")
class OrderWorkflow:

    def __init__(self) -> None:
        self.progress = 0
        self.retry_policy = OrderActivities.retry_policy

    @workflow.run
    async def execute(self, input: OrderInput) -> OrderOutput:
        workflow_type = workflow.info().workflow_type
        workflow.logger.info(f"Order workflow started, type = {workflow_type}, orderId = {input.OrderId}")

        # Get items
        order_items = await workflow.execute_local_activity_method(
            OrderActivities.get_items,
            start_to_close_timeout=timedelta(seconds=5)
        )

        # Check fraud
        await workflow.execute_activity_method(
            OrderActivities.check_fraud,
            input,
            start_to_close_timeout=timedelta(seconds=5),
            retry_policy=self.retry_policy
        )
        await self.sleep(1, 25)

        # Prepare shipment
        await workflow.execute_activity_method(
            OrderActivities.prepare_shipment,
            input,
            start_to_close_timeout=timedelta(seconds=5),
            retry_policy=self.retry_policy
        )
        await self.sleep(1, 50)

        # Charge customer
        await workflow.execute_activity_method(
            OrderActivities.charge_customer,
            args=[input, workflow_type],
            start_to_close_timeout=timedelta(seconds=5),
            retry_policy=self.retry_policy
        )
        await self.sleep(3, 75)

        # Ship order items
        handles = []
        for item in order_items:
            workflow.logger.info(f"Shipping item: {item.description}")
            handles.append(
                workflow.execute_activity_method(
                    OrderActivities.ship_order,
                    args=[input, item],
                    start_to_close_timeout=timedelta(seconds=5),
                    retry_policy=self.retry_policy
                )
            )

        # Wait for all items to ship
        await asyncio.gather(*handles)
        await self.sleep(0, 100)

        # Generate trackingId
        tracking_id = str(workflow.uuid4())
        return OrderOutput(tracking_id, input.Address)

    @workflow.query(name="getProgress")
    def query_progress(self) -> int:
        return self.progress

    async def sleep(self, seconds: int, progress: int):
        if seconds > 0:
            await asyncio.sleep(seconds)
        self.progress = progress

File: dotnet/src/OrderWorkflow.cs
using System.Collections;
using DotNetOrderManagement.model;
using Microsoft.Extensions.Logging;
using Temporalio.Workflows;

namespace DotNetOrderManagement;

[Workflow("OrderWorkflowHappyPath")]
public class OrderWorkflow
{
    private int _progress = 0;
    
    [WorkflowRun]
    public async Task<OrderOutput> Execute(OrderInput input)
    {
        var logger = Workflow.Logger;
        var type = Workflow.Info.WorkflowType;
        logger.LogInformation("Order workflow started, type = {}, orderId = {}", type, input.OrderId);
        
        // Get Items
        var orderItems = await Workflow.ExecuteLocalActivityAsync((OrderActivities act) => 
            act.GetItems(), OrderActivities.LocalActivityOpts);
        
        // Check fraud
        await Workflow.ExecuteActivityAsync((OrderActivities act) =>
            act.CheckFraud(input), OrderActivities.ActivityOpts);
        await Sleep(1, 25);
        
        // Prepare shipment
        await Workflow.ExecuteActivityAsync((OrderActivities act) => 
            act.PrepareShipment(input), OrderActivities.ActivityOpts);
        await Sleep(1, 50);
        
        // Charge Customer
        await Workflow.ExecuteActivityAsync((OrderActivities act) => 
            act.ChargeCustomer(input, type), OrderActivities.ActivityOpts);
        await Sleep(3, 75);
        
        // Ship Order
        var activities = new List<Task>();
        foreach (var orderItem in orderItems)
        {
            logger.LogInformation("Shipping item: {}", orderItem.Description);
            activities.Add(
                Workflow.ExecuteActivityAsync((OrderActivities act) => 
                    act.ShipOrder(input, orderItem), OrderActivities.ActivityOpts)
            );
        }
        
        // Wait for all items to ship
        await Workflow.WhenAllAsync(activities);
        await Sleep(1, 100);
        
        // Generate tracking ID
        var trackingId = Workflow.Random.Next().ToString();
        return new OrderOutput(trackingId, input.Address);
    }

    [WorkflowQuery("getProgress")]
    public int QueryProgress()
    {
        return _progress;
    }

    private async Task Sleep(int sleep, int progress)
    {
        _progress = progress;
        if (sleep > 0)
        {
            await Workflow.DelayAsync(TimeSpan.FromSeconds(sleep));
        }
    }
}